text/0000-time-improvements.md

```diff
@@ -0,0 +1,330 @@
+# Summary
+
+This RFC proposes several new types and associated APIs for working with times in Rust.
+The primary new types are `ProcessTime`, for working with monotonic time within a single
+process, and `SystemTime`, for working with times across processes on a single system
+(usually internally represented as a number of seconds since an epoch).
+
+# Motivations
+
+The primary motivation of this RFC is to flesh out a larger set of APIs for
+representing instants in time and durations of time.
+
+For various reasons that this RFC will explore, APIs related to time are fairly
+error-prone and have a number of caveats that programmers do not expect.
+
+Rust APIs tend to expose more of these kinds of caveats through their APIs, in
+order to help programmers become aware of and handle edge-cases. At the same
+time, un-ergonomic APIs can work against that goal.
+
+This RFC attempts to balance the desire to expose common footguns and help
+programmers handle edge-cases with a desire to avoid creating so many hoops to
+jump through that the useful caveats get ignored.
+
+At a high level, this RFC covers two concepts related to time:
+
+* Instants, moments in time
+* Durations, an amount of time between two instants
+
+We would like to be able to do some basic operations with these instants:
+
+* Compare two instants
+* Add a time period to an instant
+* Subtract a time period from an instant
+* Compare an instant to "now" to discover time elapsed
+
+However, there are a number of problems that arise when trying to define these
+types and operations.
+
+First of all, with the exception of instants created in the rutime of a single
+process, instants are not monotonic. A simple example of this is that if a
+program creates two files sequentially, it cannot assume that the creation time
+of the second file is later than the creation time of the first file.
+
+This is because NTP (the network time protocol) can arbitrarily change the
+system clock, and can even **rewind time**. This kind of time travel means that
+the "system time-line" is not continuous and monotonic, which is something that
+programmers very often forget when writing code involving machine times.
+
+This design attempts to help programmers avoid some of the most egregious and
+unexpected consequences of this kind of "time travel".
+
+---
+
+Leap seconds, which cannot be predicted, mean that it is impossible
+to reliably add a number of seconds to a particular instant represented as a
+human date and time ("1 million seconds from 2015-09-20 at midnight").
+
+They also mean that seemingly simple concepts, like "1 minute", have caveats
+depending on exactly how they are used. Caveats related to leap seconds
+create real-world bugs, because of how unusual leap seconds are, and how
+unlikely programmers are to consider "12:00:60" as a valid time.
+
+Certain kinds of seemingly simple operations may not make sense in
+all cases. For example, adding "1 year" to February 29, 2012 would produce
```

Comment:
I'd replace `would` with `could` or `might`, unless we intend to have the Rust api generate invalid dates. (The subtext here is that there are other ways to resolve these than generating invalid dates)